### PR TITLE
Workload identity

### LINKED ISSUE #334: Add support for Workload Identity as authentication mechanism
**What would you like to be added**:
Add support for Workload Identity as authentication mechanism

**Why is this needed**:

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
**What this PR does / why we need it**:
Add support for OKE workload identity

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #334

### CODE REVIEW

**Comment by shyamradhakrishnan** [comment]
unit test
```
?   	github.com/oracle/cluster-api-provider-oci	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta1	60.896s	coverage: 23.7% of statements
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta2	0.040s	coverage: 15.6% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/config	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/metrics	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/ociutil	0.009s	coverage: 19.6% of statements
ok  	github.com/oracle/cluster-api-provider-oci/cloud/scope	232.210s	coverage: 75.1% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer/mock_lb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/util	0.223s	coverage: 61.1% of statements
ok  	github.com/oracle/cluster-api-provider-oci/controllers	26.892s	coverage: 57.9% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1	15.980s	coverage: 15.2% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta2	0.017s	coverage: 4.8% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/controllers	0.942s	coverage: 56.0% of statements
?   	github.com/oracle/cluster-api-provider-oci/feature	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/version	[no test files]
go tool cover -func=coverage.out -o coverage.txt
go tool cover -html=coverage.out -o coverage.html
```

File: cloud/util/util.go [hunk 1]
```diff
@@ -20,7 +20,12 @@ import (
 	"context"
 	"crypto/x509"
 	"fmt"
+	"io"
+	"net/http"
+	"os"
 	"reflect"
+	"sigs.k8s.io/controller-runtime/pkg/log"
```
Comment (joekr) [inline cloud/util/util.go:27]: NIT: move to 3rd party packages list

File: cloud/util/util.go [hunk 2]
```diff
@@ -155,6 +169,42 @@ func GetOrBuildClientFromIdentity(ctx context.Context, c client.Client, identity
...
 			ClientOverrides:       clientOverrides})
 
+		if err != nil {
+			return nil, err
+		}
+		return clientProvider, nil
+	} else if identity.Spec.Type == infrastructurev1beta2.WorkloadPrincipal {
+		_, containsVersion := os.LookupEnv(auth.ResourcePrincipalVersionEnvVar)
+		if !containsVersion {
+			os.Setenv(auth.ResourcePrincipalVersionEnvVar, auth.ResourcePrincipalVersion1_1)
```
Comment (joekr) [inline cloud/util/util.go:179]: I don't know enough here, why 1.1 and not 2.2 for the ResourcePrincipalVersion?

**Review by joekr — APPROVED** [review]

File: cloud/util/util.go (hunk 1 above)
Comment (shyamradhakrishnan) [reply cloud/util/util.go]: fixed

File: cloud/util/util.go (hunk 2 above)
Comment (shyamradhakrishnan) [reply cloud/util/util.go]: yeah, this was for testing, it should be 2.2

**Review by joekr — APPROVED** [review]

**Comment by shyamradhakrishnan** [comment]
e2e 
```
ynchronizedAfterSuite] PASSED [1.062 seconds]
[SynchronizedAfterSuite]
/home/ubuntu/go/src/github.com/oracle/cluster-api-provider-oci/test/e2e/e2e_suite_test.go:261

  Timeline >>
  STEP: Tearing down the management cluster @ 10/03/23 04:32:44.376
  << Timeline
------------------------------

Ran 7 of 28 Specs in 2801.695 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 21 Skipped
```

**Pull request merged**